crates/sui-core/src/authority.rs

```diff
@@ -1694,16 +1694,28 @@ impl AuthorityState {
         Ok(Some(match type_ {
             DynamicFieldType::DynamicObject => {
                 // Find the actual object from storage using the object id obtained from the wrapper.
-                let Some(object) = self.database.find_object_lt_or_eq_version(object_id, o.version()) else{
-                    return Err(UserInputError::ObjectNotFound {
-                        object_id,
-                        version: Some(o.version()),
-                    }.into())
-                };
-                let version = object.version();
-                let digest = object.digest();
-                let object_type = object.data.type_().unwrap();
 
+                // Try to find the object in the written objects first.
+                let (version, digest, object_type) =
+                    if let Some((_, object, _)) = written.get(&object_id) {
+                        let version = object.version();
+                        let digest = object.digest();
+                        let object_type = object.data.type_().unwrap().clone();
+                        (version, digest, object_type)
+                    } else {
+                        // If not found, try to find it in the database.
```

Comment:
Do we need this? I suspect all the dynamic field objects needed to be index should be in WrittenObjects already.

@longbowlu, @amnn, @666lcz WDYT?

Reply:
How is this object different from the object passed into this function (line 1663)?

Reply:
the object in line 1663 is the dynamic field wrapper, this logic try to retrieve the actual object wrapped by the wrapper.